File: 95. Unique Binary Search Trees II/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func generateTrees(n int) []*TreeNode {
	memo = make([][][]*TreeNode, n)
	for i := 0; i < n; i++ {
		memo[i] = make([][]*TreeNode, n)
	}

	return getTrees(1, n)
}

var memo [][][]*TreeNode

func getTrees(lo int, hi int) []*TreeNode {
	if lo > hi {
		return []*TreeNode{nil}
	}

	if len(memo[lo-1][hi-1]) != 0 {
		return memo[lo-1][hi-1]
	}

	res := []*TreeNode{}
	for i := lo; i <= hi; i++ {

		left := getTrees(lo, i-1)
		right := getTrees(i+1, hi)

		for _, l := range left {
			for _, r := range right {
				res = append(res, &TreeNode{
					Val:   i,
					Left:  l,
					Right: r,
				})
			}
		}
	}

	memo[lo-1][hi-1] = res
	return res
}

File: 95. Unique Binary Search Trees II/main.py
# Definition for a binary tree node.
from distutils.command.build import build
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:

        def gen(lo: int, hi: int) -> List[TreeNode]:
            res = []

            if lo > hi:
                res.append(None)
                return res

            for i in range(lo, hi + 1):
                l = gen(lo, i-1)
                r = gen(i+1, hi)

                for left in l:
                    for right in r:
                        root = TreeNode(i)
                        root.left = left
                        root.right = right
                        res.append(root)

            return res
        
        if n == 0:
            return []
        return build(1, n)